File: mimi-app/src/main/java/com/emogoth/android/phone/mimi/db/CatalogTableConnection.java
package com.emogoth.android.phone.mimi.db;


import com.emogoth.android.phone.mimi.db.model.CatalogPostModel;
import com.mimireader.chanlib.models.ChanCatalog;
import com.mimireader.chanlib.models.ChanPost;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import io.reactivex.Flowable;
import io.reactivex.Observable;
import io.reactivex.Single;
import io.reactivex.functions.Function;

public class CatalogTableConnection {
    public static final String LOG_TAG = CatalogTableConnection.class.getSimpleName();

    public static Single<List<CatalogPostModel>> fetchPosts() {
        return DatabaseUtils.fetchTable(CatalogPostModel.class, CatalogPostModel.TABLE_NAME);
    }

    public static Function<List<CatalogPostModel>, List<ChanPost>> convertDbPostsToChanPosts() {
        return CatalogPostModels -> {
            if (CatalogPostModels == null || CatalogPostModels.size() == 0) {
                return Collections.emptyList();
            }

            List<ChanPost> posts = new ArrayList<>(CatalogPostModels.size());
            for (CatalogPostModel dbPost : CatalogPostModels) {
                posts.add(dbPost.toPost());
            }

            return posts;
        };
    }

    public static Observable<Boolean> putPosts(ChanCatalog catalog) {
        if (catalog == null) {
            return Observable.just(false);
        }

        List<ChanPost> posts = catalog.getPosts();

        if (posts == null) {
            return Observable.just(false);
        }

        List<CatalogPostModel> catalogPosts = new ArrayList<>(posts.size());
        for (int i = 0; i < posts.size(); i++) {
            catalogPosts.add(new CatalogPostModel(posts.get(i)));
        }

        return DatabaseUtils.insert(catalogPosts).first(false).toObservable();
    }

    public static Single<Boolean> removeThread(long threadId) {
        CatalogPostModel model = new CatalogPostModel();
        DatabaseUtils.WhereArg[] args = new DatabaseUtils.WhereArg[1];

        args[0] = new DatabaseUtils.WhereArg(CatalogPostModel.KEY_POST_ID + "=?", threadId);
        return DatabaseUtils.remove(model, true, args).single(false);
    }

    public static Single<Boolean> clear() {
        CatalogPostModel model = new CatalogPostModel();
        DatabaseUtils.WhereArg[] args = new DatabaseUtils.WhereArg[0];
        return DatabaseUtils.remove(model, true, args).single(false);
    }
}

File: mimi-app/src/main/java/com/emogoth/android/phone/mimi/db/CatalogTableConnection.kt
package com.emogoth.android.phone.mimi.db

import com.emogoth.android.phone.mimi.db.MimiDatabase.Companion.getInstance
import com.emogoth.android.phone.mimi.db.models.CatalogPost
import com.mimireader.chanlib.models.ChanCatalog
import com.mimireader.chanlib.models.ChanPost
import io.reactivex.Single
import io.reactivex.functions.Function

object CatalogTableConnection {
    val LOG_TAG = CatalogTableConnection::class.java.simpleName
    @JvmStatic
    fun fetchPosts(): Single<List<CatalogPost>> {
        return getInstance()!!.catalog().getAll().firstOrError()
    }

    @JvmStatic
    fun convertDbPostsToChanPosts(): Function<List<CatalogPost>, List<ChanPost>> {
        return Function { CatalogPostModels: List<CatalogPost> ->
            val posts: ArrayList<ChanPost> = ArrayList(CatalogPostModels.size)
            for (dbPost in CatalogPostModels) {
                posts.add(dbPost.toPost())
            }
            posts
        }
    }

    @JvmStatic
    fun putPosts(catalog: ChanCatalog): Single<Boolean> {
        return Single.defer {
            val posts = catalog.posts ?: return@defer Single.just(false)
            val catalogPosts: MutableList<CatalogPost> = ArrayList(posts.size)
            for (i in posts.indices) {
                catalogPosts.add(CatalogPost(posts[i]))
            }
            val resultsList = getInstance()!!.catalog().insert(catalogPosts)
            var success = true
            for (value in resultsList) {
                if (success) {
                    success = value > 0
                }
            }
            Single.just(success)
        }
    }

    @JvmStatic
    fun removeThread(threadId: Long): Single<Boolean> {
        return Single.defer {
            getInstance()?.catalog()?.removeThread(threadId) ?: return@defer Single.just(false)
            Single.just(true)
        }
    }

    @JvmStatic
    fun clear(): Single<Boolean> {
        return Single.defer {
            getInstance()?.catalog()?.clear() ?: return@defer Single.just(false)
            Single.just(true)
        }
    }
}